File: ocaml/lib/jsonrpc.ml
(* Copyright (c) Microsoft Corporation. All rights reserved. *)

(** JSON-RPC 2.0 transport layer with Content-Length header framing. *)

(* ========================================================================== *)
(* Error types                                                                *)
(* ========================================================================== *)

type jsonrpc_error = {
  code : int;
  message : string;
  data : Yojson.Safe.t option;
}

exception Jsonrpc_error of jsonrpc_error
exception Process_exited of string

type request_handler = string -> Yojson.Safe.t -> Yojson.Safe.t Lwt.t

(* ========================================================================== *)
(* Internal state                                                             *)
(* ========================================================================== *)

type t = {
  ic : Lwt_io.input_channel;
  oc : Lwt_io.output_channel;
  mutable next_id : int;
  pending : (int, Yojson.Safe.t Lwt.u) Hashtbl.t;
  handlers : (string, request_handler) Hashtbl.t;
  mutable notification_handler :
    (string -> Yojson.Safe.t -> unit Lwt.t) option;
  mutable running : bool;
  write_mutex : Lwt_mutex.t;
}

(* ========================================================================== *)
(* Construction                                                               *)
(* ========================================================================== *)

let create ic oc =
  { ic
  ; oc
  ; next_id = 1
  ; pending = Hashtbl.create 16
  ; handlers = Hashtbl.create 8
  ; notification_handler = None
  ; running = false
  ; write_mutex = Lwt_mutex.create ()
  }

(* ========================================================================== *)
(* Wire format: Content-Length header framing                                  *)
(* ========================================================================== *)

let read_message (ic : Lwt_io.input_channel) : Yojson.Safe.t option Lwt.t =
  let open Lwt.Syntax in
  let rec read_headers content_length =
    let* line = Lwt_io.read_line_opt ic in
    match line with
    | None -> Lwt.return_none
    | Some "" | Some "\r" ->
      (* End of headers *)
      (match content_length with
       | None -> Lwt.return_none
       | Some len ->
         let buf = Bytes.create len in
         let* () = Lwt_io.read_into_exactly ic buf 0 len in
         let body = Bytes.to_string buf in
         let json = Yojson.Safe.from_string body in
         Lwt.return_some json)
    | Some header ->
      let header = String.trim header in
      let cl =
        if String.length header > 16
           && String.lowercase_ascii (String.sub header 0 16)
              = "content-length: "
        then
          Some (int_of_string (String.trim (String.sub header 16
                 (String.length header - 16))))
        else content_length
      in
      read_headers cl
  in
  read_headers None

let write_message (oc : Lwt_io.output_channel) (mutex : Lwt_mutex.t)
    (json : Yojson.Safe.t) : unit Lwt.t =
  let open Lwt.Syntax in
  let body = Yojson.Safe.to_string json in
  let header =
    Printf.sprintf "Content-Length: %d\r\n\r\n" (String.length body)
  in
  Lwt_mutex.with_lock mutex (fun () ->
    let* () = Lwt_io.write oc header in
    let* () = Lwt_io.write oc body in
    Lwt_io.flush oc)

(* ========================================================================== *)
(* Message reading loop                                                       *)
(* ========================================================================== *)

let handle_response (t : t) (id : int) (result : Yojson.Safe.t option)
    (error : Yojson.Safe.t option) : unit =
  match Hashtbl.find_opt t.pending id with
  | None -> ()
  | Some resolver ->
    Hashtbl.remove t.pending id;
    (match error with
     | Some err ->
       let open Yojson.Safe.Util in
       let code = try err |> member "code" |> to_int with _ -> -1 in
       let message =
         try err |> member "message" |> to_string with _ -> "Unknown error"
       in
       let data =
         try Some (err |> member "data") with _ -> None
       in
       Lwt.wakeup_exn resolver
         (Jsonrpc_error { code; message; data })
     | None ->
       let value = match result with Some v -> v | None -> `Null in
       Lwt.wakeup resolver value)

let handle_request (t : t) (id : Yojson.Safe.t) (method_ : string)
    (params : Yojson.Safe.t) : unit Lwt.t =
  let open Lwt.Syntax in
  match Hashtbl.find_opt t.handlers method_ with
  | None ->
    let resp =
      `Assoc
        [ ("jsonrpc", `String "2.0")
        ; ("id", id)
        ; ("error",
           `Assoc
             [ ("code", `Int (-32601))
             ; ("message", `String (Printf.sprintf "Method not found: %s" method_))
             ])
        ]
    in
    write_message t.oc t.write_mutex resp
  | Some handler ->
    let* result =
      Lwt.catch
        (fun () ->
          let* r = handler method_ params in
          Lwt.return_ok r)
        (fun exn -> Lwt.return_error (Printexc.to_string exn))
    in
    let resp =
      match result with
      | Ok value ->
        `Assoc
          [ ("jsonrpc", `String "2.0")
          ; ("id", id)
          ; ("result", value)
          ]
      | Error msg ->
        `Assoc
          [ ("jsonrpc", `String "2.0")
          ; ("id", id)
          ; ("error",
             `Assoc
               [ ("code", `Int (-32603))
               ; ("message", `String msg)
               ])
          ]
    in
    write_message t.oc t.write_mutex resp

let handle_notification (t : t) (method_ : string) (params : Yojson.Safe.t)
    : unit Lwt.t =
  match t.notification_handler with
  | None -> Lwt.return_unit
  | Some handler -> handler method_ params

let read_loop (t : t) : unit Lwt.t =
  let open Lwt.Syntax in
  let rec loop () =
    if not t.running then Lwt.return_unit
    else
      let* msg = read_message t.ic in
      match msg with
      | None ->
        (* EOF -- mark pending requests as failed *)
        t.running <- false;
        Hashtbl.iter
          (fun _id resolver ->
            Lwt.wakeup_exn resolver (Process_exited "CLI process exited"))
          t.pending;
        Hashtbl.reset t.pending;
        Lwt.return_unit
      | Some json ->
        let open Yojson.Safe.Util in
        let id_opt =
          try Some (json |> member "id") with _ -> None
        in
        let method_opt =
          try Some (json |> member "method" |> to_string) with _ -> None
        in
        let result_opt =
          try Some (json |> member "result") with _ -> None
        in
        let error_opt =
          try
            let e = json |> member "error" in
            if e = `Null then None else Some e
          with _ -> None
        in
        let* () =
          match id_opt, method_opt with
          | Some (`Int id), None ->
            (* Response to our request *)
            handle_response t id result_opt error_opt;
            Lwt.return_unit
          | Some id, Some method_ ->
            (* Incoming request *)
            handle_request t id method_ (try json |> member "params" with _ -> `Null)
          | None, Some method_ ->
            (* Incoming notification *)
            handle_notification t method_ (try json |> member "params" with _ -> `Null)
          | _ -> Lwt.return_unit
        in
        loop ()
  in
  loop ()

(* ========================================================================== *)
(* Public API                                                                 *)
(* ========================================================================== *)

let start (t : t) : unit Lwt.t =
  t.running <- true;
  Lwt.async (fun () ->
    Lwt.catch (fun () -> read_loop t) (fun _exn ->
      t.running <- false;
      Lwt.return_unit));
  Lwt.return_unit

let stop (t : t) : unit Lwt.t =
  let open Lwt.Syntax in
  t.running <- false;
  Hashtbl.iter
    (fun _id resolver ->
      Lwt.wakeup_exn resolver (Process_exited "Client stopped"))
    t.pending;
  Hashtbl.reset t.pending;
  let* () =
    Lwt.catch (fun () -> Lwt_io.close t.oc) (fun _ -> Lwt.return_unit)
  in
  Lwt.catch (fun () -> Lwt_io.close t.ic) (fun _ -> Lwt.return_unit)

let send_request (t : t) (method_ : string) (params : Yojson.Safe.t)
    : Yojson.Safe.t Lwt.t =
  if not t.running then
    Lwt.fail (Process_exited "Client is not running");
  let id = t.next_id in
  t.next_id <- t.next_id + 1;
  let promise, resolver = Lwt.wait () in
  Hashtbl.add t.pending id resolver;
  let msg =
    `Assoc
      [ ("jsonrpc", `String "2.0")
      ; ("id", `Int id)
      ; ("method", `String method_)
      ; ("params", params)
      ]
  in
  let open Lwt.Syntax in
  let* () = write_message t.oc t.write_mutex msg in
  promise

let send_notification (t : t) (method_ : string) (params : Yojson.Safe.t)
    : unit Lwt.t =
  let msg =
    `Assoc
      [ ("jsonrpc", `String "2.0")
      ; ("method", `String method_)
      ; ("params", params)
      ]
  in
  write_message t.oc t.write_mutex msg

let register_handler (t : t) (method_ : string) (handler : request_handler)
    : unit =
  Hashtbl.replace t.handlers method_ handler

let set_notification_handler (t : t)
    (handler : string -> Yojson.Safe.t -> unit Lwt.t) : unit =
  t.notification_handler <- Some handler

let is_running (t : t) : bool = t.running

File: fsharp/src/JsonRpc.fs
namespace CopilotSDK.Supercharged.FSharp

open System
open System.Buffers
open System.IO
open System.Text
open System.Text.Json
open System.Threading
open System.Threading.Tasks

// ---------------------------------------------------------------------------
// JSON-RPC 2.0 transport with Content-Length header framing (LSP style)
// ---------------------------------------------------------------------------

/// A JSON-RPC 2.0 request envelope.
type JsonRpcRequest =
    { [<System.Text.Json.Serialization.JsonPropertyName("jsonrpc")>]
      JsonRpc: string
      [<System.Text.Json.Serialization.JsonPropertyName("id")>]
      Id: int
      [<System.Text.Json.Serialization.JsonPropertyName("method")>]
      Method: string
      [<System.Text.Json.Serialization.JsonPropertyName("params")>]
      Params: obj option }

/// A JSON-RPC 2.0 response envelope.
type JsonRpcResponse =
    { [<System.Text.Json.Serialization.JsonPropertyName("jsonrpc")>]
      JsonRpc: string
      [<System.Text.Json.Serialization.JsonPropertyName("id")>]
      Id: int option
      [<System.Text.Json.Serialization.JsonPropertyName("result")>]
      Result: JsonElement option
      [<System.Text.Json.Serialization.JsonPropertyName("error")>]
      Error: JsonRpcError option }

/// A JSON-RPC 2.0 error object.
and JsonRpcError =
    { [<System.Text.Json.Serialization.JsonPropertyName("code")>]
      Code: int
      [<System.Text.Json.Serialization.JsonPropertyName("message")>]
      Message: string
      [<System.Text.Json.Serialization.JsonPropertyName("data")>]
      Data: JsonElement option }

/// A JSON-RPC 2.0 notification (no id, no response expected).
type JsonRpcNotification =
    { [<System.Text.Json.Serialization.JsonPropertyName("jsonrpc")>]
      JsonRpc: string
      [<System.Text.Json.Serialization.JsonPropertyName("method")>]
      Method: string
      [<System.Text.Json.Serialization.JsonPropertyName("params")>]
      Params: obj option }

/// Exception raised when a JSON-RPC error response is received.
type JsonRpcException(code: int, message: string, data: JsonElement option) =
    inherit Exception(sprintf "JSON-RPC error %d: %s" code message)
    member _.Code = code
    member _.ErrorData = data

/// Low-level JSON-RPC transport over streams with Content-Length framing.
type JsonRpcTransport(input: Stream, output: Stream) =
    let jsonOpts = JsonSerializerOptions(PropertyNamingPolicy = JsonNamingPolicy.CamelCase)
    let mutable nextId = 0
    let pendingRequests = System.Collections.Concurrent.ConcurrentDictionary<int, TaskCompletionSource<JsonRpcResponse>>()
    let notificationHandlers = System.Collections.Concurrent.ConcurrentDictionary<string, Action<JsonElement option>>()
    let requestHandlers = System.Collections.Concurrent.ConcurrentDictionary<string, JsonElement option -> Async<obj>>()
    let writeLock = new SemaphoreSlim(1, 1)
    let cts = new CancellationTokenSource()

    let readHeaderLine (reader: StreamReader) = async {
        let! line = reader.ReadLineAsync() |> Async.AwaitTask
        return line
    }

    let readMessage (reader: StreamReader) = async {
        let mutable contentLength = -1
        let mutable keepReading = true

        while keepReading do
            let! line = readHeaderLine reader
            if isNull line then
                keepReading <- false
            elif line = "" then
                keepReading <- false
            elif line.StartsWith("Content-Length:", StringComparison.OrdinalIgnoreCase) then
                let value = line.Substring("Content-Length:".Length).Trim()
                contentLength <- Int32.Parse(value)
            else
                ()

        if contentLength <= 0 then
            return None
        else
            let buffer = Array.zeroCreate<char> contentLength
            let mutable totalRead = 0
            while totalRead < contentLength do
                let! read = reader.ReadAsync(buffer, totalRead, contentLength - totalRead) |> Async.AwaitTask
                if read = 0 then
                    failwith "Unexpected end of stream"
                totalRead <- totalRead + read
            let json = new string(buffer, 0, contentLength)
            return Some json
    }

    let writeMessage (json: string) = async {
        let bytes = Encoding.UTF8.GetBytes(json)
        let header = sprintf "Content-Length: %d\r\n\r\n" bytes.Length
        let headerBytes = Encoding.UTF8.GetBytes(header)
        do! writeLock.WaitAsync() |> Async.AwaitTask
        try
            do! output.WriteAsync(headerBytes, 0, headerBytes.Length) |> Async.AwaitTask
            do! output.WriteAsync(bytes, 0, bytes.Length) |> Async.AwaitTask
            do! output.FlushAsync() |> Async.AwaitTask
        finally
            writeLock.Release() |> ignore
    }

    let sendJsonRpcResponse (idRawText: string) (result: obj) = async {
        let resultJson = JsonSerializer.Serialize(result, jsonOpts)
        let responseMsg = sprintf """{"jsonrpc":"2.0","id":%s,"result":%s}""" idRawText resultJson
        do! writeMessage responseMsg
    }

    let sendJsonRpcErrorResponse (idRawText: string) (code: int) (errMessage: string) = async {
        let escapedMsg = JsonSerializer.Serialize(errMessage)
        let responseMsg = sprintf """{"jsonrpc":"2.0","id":%s,"error":{"code":%d,"message":%s}}""" idRawText code escapedMsg
        do! writeMessage responseMsg
    }

    /// Dispatch an incoming JSON message to pending requests or notification handlers.
    let dispatchMessage (json: string) =
        let doc = JsonDocument.Parse(json)
        let root = doc.RootElement

        // Check if this is a response (has "id" and either "result" or "error")
        let hasId =
            match root.TryGetProperty("id") with
            | true, idElem when idElem.ValueKind <> JsonValueKind.Null -> true
            | _ -> false

        let hasResult =
            match root.TryGetProperty("result") with
            | true, _ -> true
            | _ -> false

        let hasError =
            match root.TryGetProperty("error") with
            | true, _ -> true
            | _ -> false

        if hasId && (hasResult || hasError) then
            // It is a response
            let response = JsonSerializer.Deserialize<JsonRpcResponse>(json, jsonOpts)
            match response.Id with
            | Some id ->
                match pendingRequests.TryRemove(id) with
                | true, tcs -> tcs.SetResult(response)
                | _ -> ()
            | None -> ()
        else
            // It is a notification or a server-initiated request
            match root.TryGetProperty("method") with
            | true, methodElem ->
                let methodName = methodElem.GetString()
                let paramData =
                    match root.TryGetProperty("params") with
                    | true, p -> Some (p.Clone())
                    | _ -> None
                if hasId then
                    // Server-initiated request — dispatch and send response
                    let idRawText =
                        match root.TryGetProperty("id") with
                        | true, idElem -> idElem.GetRawText()
                        | _ -> "null"
                    match requestHandlers.TryGetValue(methodName) with
                    | true, handler ->
                        async {
                            try
                                let! result = handler paramData
                                do! sendJsonRpcResponse idRawText result
                            with ex ->
                                do! sendJsonRpcErrorResponse idRawText -32000 ex.Message
                        } |> Async.Start
                    | _ ->
                        match notificationHandlers.TryGetValue(methodName) with
                        | true, handler -> handler.Invoke(paramData)
                        | _ -> ()
                else
                    match notificationHandlers.TryGetValue(methodName) with
                    | true, handler -> handler.Invoke(paramData)
                    | _ -> ()
            | _ -> ()

    /// Start the read loop that dispatches incoming messages.
    member _.StartReadLoop() =
        let reader = new StreamReader(input, Encoding.UTF8)
        let loop = async {
            try
                let mutable running = true
                while running && not cts.Token.IsCancellationRequested do
                    let! msg = readMessage reader
                    match msg with
                    | Some json -> dispatchMessage json
                    | None -> running <- false
            with
            | :? OperationCanceledException -> ()
            | ex ->
                // Signal all pending requests
                for kvp in pendingRequests do
                    kvp.Value.TrySetException(ex) |> ignore
        }
        Async.Start(loop, cts.Token)

    /// Send a JSON-RPC request and wait for the response.
    member _.SendRequestAsync<'T>(method: string, ?parameters: obj) : Async<'T> = async {
        let id = Interlocked.Increment(&nextId)
        let request =
            { JsonRpc = "2.0"
              Id = id
              Method = method
              Params = parameters }
        let tcs = TaskCompletionSource<JsonRpcResponse>()
        pendingRequests.TryAdd(id, tcs) |> ignore

        let json = JsonSerializer.Serialize(request, jsonOpts)
        do! writeMessage json

        let! response = tcs.Task |> Async.AwaitTask

        match response.Error with
        | Some err ->
            return raise (JsonRpcException(err.Code, err.Message, err.Data))
        | None ->
            match response.Result with
            | Some result ->
                let opts = JsonSerializerOptions(PropertyNameCaseInsensitive = true)
                return JsonSerializer.Deserialize<'T>(result.GetRawText(), opts)
            | None ->
                return Unchecked.defaultof<'T>
    }

    /// Send a JSON-RPC notification (no response expected).
    member _.SendNotificationAsync(method: string, ?parameters: obj) : Async<unit> = async {
        let notification =
            { JsonRpc = "2.0"
              Method = method
              Params = parameters }
        let json = JsonSerializer.Serialize(notification, jsonOpts)
        do! writeMessage json
    }

    /// Register a handler for a server-initiated notification method.
    member _.OnNotification(method: string, handler: JsonElement option -> unit) =
        notificationHandlers.[method] <- Action<JsonElement option>(handler)

    /// Remove a previously registered notification handler.
    member _.RemoveNotification(method: string) =
        notificationHandlers.TryRemove(method) |> ignore

    /// Register a handler for a server-initiated request method.
    /// The handler receives params and returns a result sent back as the JSON-RPC response.
    member _.OnRequest(method: string, handler: JsonElement option -> Async<obj>) =
        requestHandlers.[method] <- handler

    /// Remove a previously registered request handler.
    member _.RemoveRequest(method: string) =
        requestHandlers.TryRemove(method) |> ignore

    /// Shut down the transport, cancelling any pending reads.
    member _.Dispose() =
        cts.Cancel()
        cts.Dispose()

    interface IDisposable with
        member this.Dispose() = this.Dispose()
